pkg/patch/patch.go

```diff
@@ -47,85 +45,87 @@ type TemplateContextRegex struct {
 func Apply(repoDir string, patch config.PackageRepoPatch, tmplCtx TemplateContext) error {
-	log.Debug().Str("file", patch.File).Msg("Patching file.")
+	fstore := NewCachedFileStore(repoDir)
+	defer fstore.Close()
+	switch {
+	case patch.Regex != nil:
+		if err := applyRegexPatch(fstore, tmplCtx, *patch.Regex); err != nil {
+			return fmt.Errorf("regex patch: %w", err)
+		}
+	case patch.YAML != nil:
+		if err := applyYAMLPatch(fstore, tmplCtx, *patch.YAML); err != nil {
+			return fmt.Errorf("yaml patch: %w", err)
+		}
+	case patch.HelmDepUpdate != nil:
+		// Flush the store as we need the up-to-date changes on disk
+		if err := fstore.Flush(); err != nil {
+			return err
+		}
+		if err := applyHelmDepUpdatePatch(repoDir, tmplCtx, *patch.HelmDepUpdate); err != nil {
+			return fmt.Errorf("exec patch: %w", err)
+		}
+	default:
+		return errors.New("missing patch type config")
+	}
 
-	// TODO: Check that the patch path doesn't go outside the repo dir.
-	// For example, reject stuff like "../../../somefile.txt"
-	path := filepath.Join(repoDir, patch.File)
+	return fstore.Close()
+}
 
-	lines, err := readLines(path)
-	if err != nil {
-		return fmt.Errorf("read file for patch: %w", err)
-	}
+func applyRegexPatch(fstore FileStore, tmplCtx TemplateContext, patch config.PatchRegex) error {
+	log.Debug().Str("file", patch.File).Stringer("match", patch.Match).Msg("Patching regex.")
 
-	if err := patchLines(patch, tmplCtx, lines); err != nil {
-		return fmt.Errorf("patch lines: %w", err)
+	if patch.File == "" {
+		return fmt.Errorf("missing required field 'file'")
 	}
-
-	if err := writeLines(path, lines); err != nil {
-		return fmt.Errorf("write patch: %w", err)
+	if patch.Match == nil {
+		return fmt.Errorf("missing required field 'match'")
+	}
+	if patch.Replace == nil {
+		return fmt.Errorf("missing required field 'replace'")
 	}
 
-	return nil
-}
+	content, err := fstore.ReadFile(patch.File)
+	if err != nil {
+		return err
+	}
+	regex := patch.Match.Regexp()
+	lines := bytes.Split(content, []byte("\n"))
 
-func patchLines(patch config.PackageRepoPatch, tmplCtx TemplateContext, lines [][]byte) error {
 	for i, line := range lines {
-		newLine, err := patchSingleLine(patch, tmplCtx, line)
-		if err != nil {
-			return err
-		}
-		if newLine == nil { // No match
+		groupIndices := regex.FindSubmatchIndex(line)
+		if groupIndices == nil {
+			// No match
 			continue
 		}
-		if bytes.Equal(line, newLine) {
-			return errors.New("found matching line, but already up-to-date")
-		}
-		lines[i] = newLine
-		return nil // Stop after first match
-	}
-	return errors.New("no match in file")
-}
 
-func patchSingleLine(patch config.PackageRepoPatch, tmplCtx TemplateContext, line []byte) ([]byte, error) {
-	switch {
-	case patch.Regex != nil:
-		return patchSingleLineRegex(*patch.Regex, tmplCtx, line)
-	default:
-		return nil, errors.New("missing patch type config")
-	}
-}
+		fullMatchStart := groupIndices[0]
+		fullMatchEnd := groupIndices[1]
```

Comment:
Are these indices guaranteed to exist?

Reply:
Yes.

- `groupIndices[0]`: start index of full match
- `groupIndices[1]`: end index of full match
- `groupIndices[2]`: start index of group 1
- `groupIndices[3]`: end index of group 1
- `groupIndices[4]`: start index of group 2
- `groupIndices[5]`: end index of group 2

Only the first two are guaranteed to exist. The other depend on how many groups you define in your regex.

The Go regex API sucks. Just tossing around some structs (or even just some arrays) in here would improve it by a landslide.